### DISCUSSION TITLE
getResultAndCount produces wrong count with join + group statements

### DISCUSSION THREAD

**Discussion by tmikoss** [mikro-orm/mikro-orm#4355] [Q&A]
**Describe the bug**
When used with join + group statements, `getResultAndCount` count query currently produces group counts for each row, rather than the total of rows.

**Stack trace**
n/a

**To Reproduce**

Given entities:

```ts
@Entity()
class Page {
  @PrimaryKey()
  id: number

  @OneToMany(() => PageVersion, (pageVersion) => pageVersion.page)
  pageVersions = new Collection<PageVersion>(this)

  @Property({ persist: false })
  pageVersionsCount?: number;
}

@Entity()
class PageVersion {
  @PrimaryKey()
  id: number
  
  @ManyToOne(() => Page)
  page: Page
}
```

and the following query builder:

```ts
    const query = em
      .createQueryBuilder(Page)
      .select(['id', 'count(pv) as "pageVersionsCount"'])
      .leftJoin('pageVersions', 'pv')
      .groupBy('id')
      .limit(1)
      .offset(1)
```

running `query.getResultAndCount()` will produce the following query for the count part:

```sql
select count(distinct("p0"."id")) as "count" from "page" as "p0" left join "page_version" as "pv" on "p0"."id" = "pv"."page_id" group by "p0"."id" 
```

With three `page` records in database, this produces the following output:

| count |
| - | 
| 1 |
| 1 |
| 1 |

**Expected behavior**

The expected query should look like this, with `group by` removed:

```sql
select count(distinct("p0"."id")) as "count" from "page" as "p0" left join "page_version" as "pv" on "p0"."id" = "pv"."page_id"
```

This would produce the following output:

| count |
| - |
| 3 |

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | v18.16.0 |
| typescript | 4.9.5 |
| mikro-orm | 5.7.3 |
| pg | 8.10.0 |

**Comment by B4nan** [comment]
Try latest version, it should disable query pagination in this case, which could help on its own.

> running query.getResultAndCount() will produce the following query for the count part:

That sounds weird, it should produce different query, with nested subqueries, as you join to-many relation and apply a limit to that. But that is what changed in latest version - it wont be applied if you provide also a `groupBy` clause.

> The expected query should look like this, with group by removed:

So you are passing there explicit `groupBy` and you expect it to be removed? Why? It is a query builder, it does what you instruct it, if you dont want the `group by` clause, dont use `groupBy()`.

**Reply by B4nan** [reply]
I think I know what you mean, and I don't think you should be using `getResultAndCount` here. Sure we could remove the `groupBy` from the count query produced by it, but that does not seem like the correct solution to me - the purpose is to return total items without limits for the same query, if you omit the group by, it will not work the same, it will return number of results without grouping, so a cartesian product. 

edit: But we use distinct count, so it could be fine? I am really not sure right now, will need to do some tests.

**Reply by B4nan** [reply]
I can take a look later, if you want to try it yourself, this is the place where we will need to reset the group by clause:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/query/QueryBuilder.ts#L681

Should be just about allowing `qb.groupBy()` without params so it can be used to reset the clause.
